### PR TITLE
Hex lattice

### PR DESCRIPTION
**Pull request by smharper**
This pull request gives OpenMC the ability to simulate geometries with hexagonal lattices.  In addition to adding the new hexagonal lattice features, this PR makes significant changes to the existing code.  Here are the top-level changes:
* Users can specify a new element, hex_lattice, in the geometry.xml file
* The documentation has been updated, and two new tests have been added.
* Lattices in the geometry.xml file no longer need the `type` element/attribute.
* `pitch` is now used rather than `width` when specifying lattices in the geometry.xml file.
* In order to handle the changes with `type` and `width`, I added deprecation warnings to input.F90.  I know that we don't technically need them since we are still in 0.x.y releases, but I think it's more polite to our users.  I also added a Python utility to automatically update geometry.xml files.

Low-level changes:
* The lattice code in geometry.F90 is now a little more modular.  Various lattice calculations were moved into the functions `get_lat_trans`, `get_lat_indices`, and `is_valid_lat_index`.
* Lattices use a polymorphic data structure---a cool feature of the Fortran 2003 standard.  There is an abstract `Lattice` class.  `Lattice` objects can be passed around by most functions without worrying about whether they are actually instances of the `RectLattice` or `HexLattice` types.  Unfortunately, it also means that the global `lattices` array needs to be accessed like `lattices(i) % obj` rather than just `lattices(i)`.  Sometime in the future I would like to convert the lattice primitives like `get_lat_trans` into type-bound procedures.
* Removed `lattice_edge` calculations from `find_cell`.  This had no effect on the test suite, and I have not seen any damage caused by it's removal in my larger test calculations.
* `distance_to_boundary` stores lattice distances and surface distances in different variables in order to address issue #258

Design decisions that may affect performance or readability:
* Hexagonal lattice cell indexing has an algorithm that uses the particle's distance from various cell centers.  It could probably be replaced with an algorithm that uses division remainders, but all of my attempted remainder algorithms suffered from finite precision issues that caused lost particles.
* The hexagonal lattice portion of `distance_to_boundary` uses a lot of expensive calculations to measure distances relative to the neighbor cells rather than the current cell.  This algorithm is also used due to it's superior finite precision performance.
* There is a check in `distance_to_boundary` for negative lattice distances.  I needed this for debugging, but I am considering leaving it in so that no future developers accidentally make changes that allow negative distances.
* `cross_lattice` now uses `get_lat_trans` which is a little bit more expensive than the older method of subtracting cell widths from the particle position, but it's agnostic to the lattice type and provides better finite precision performance (I had a lot of finite precision issues, can you tell?).

Verification and validation:
* I tested a large 2d hex lattice model against a model of manually placed cylinders.  I got a disagreement of 3 +/- 12 pcm.
* I tested a 3d hex lattice model against a model using a 2d hex lattice placed in a 3d rectangular lattice.  I got a disagreement of 17 +/- 22 pcm.
* I used an earlier version of hex_lattice to model various cores in the LR-0 reactor.  The calculations differed from MCNP5 calculations by roughly 30 +/ 10 pcm and agreed with the experiment within uncertainties.

I can provide my V&V models to anyone who wants them.  I'll also try to rerun them after we've gotten through a lot of the review process to make sure no errors have sneaked in.  Also, I should mention that none of the regression test results are different (despite the changes to the handling of rectangular lattices) with the exception of score_flux_yn.  I haven't updated that test yet because I am investigating it to see if I can find the reason for the change.

### CODE REVIEW

**Comment by smharper** [comment]
The bug that I mentioned in the comment above was fixed in #359 so I've now removed the crappy fix implemented in 11c490b (it turned out to cause another bug anyway).  I have also moved the stuff in the update_lattices.py script over to the update_inputs.py script.

At this point, I don't know of any standing issues with this PR.

File: docs/source/methods/geometry.rst [hunk 1]
```diff
@@ -385,6 +385,101 @@ is found that contains the specified point.
...
+   Rectilinear lattice tile indices.
+
+In general, a lattice tile is specified by the three indices,
+:math:`(i_x, i_y, i_z)`.  If a particle's current coordinates are
+:math:`(x, y, z)` then the indices can be determined from these formulas:
+
+.. math::
+    :label: rect_indexing
+
+    i_x = \Bigg \lceil \frac{x - x_0}{p_0} \Bigg \rceil
```
Comment (paulromano) [inline docs/source/methods/geometry.rst]: Did \left and \right (instead of \Bigg) not work for this?

File: docs/source/methods/geometry.rst [hunk 2]
```diff
@@ -385,6 +385,101 @@ is found that contains the specified point.
...
+volumetric units of the lattice will be referred to as "tiles".  Tiles are
+identified by thier indices, and the process of discovering which tile contains
+the particle is referred to as "indexing".
+
+Rectilinear Lattice Indexing
+----------------------------
+
+Indices are assigned to tiles in a rectilinear lattice based on the tile's
+position along the :math:`x`, :math:`y`, and :math:`z` axes.  The figure below
+maps the indecies for a 2d lattice.  The indices, (1, 1), map to the
```
Comment (paulromano) [inline docs/source/methods/geometry.rst]: typo on "indecies". Also, I'd prefer "2D" over "2d" (another instance further down also).

File: docs/source/methods/geometry.rst [hunk 3]
```diff
@@ -385,6 +385,101 @@ is found that contains the specified point.
...
+.. figure:: ../_images/hex_lat.*
+   :align: center
+   :figclass: align-center
+   :width: 400px
+
+   Hexagonal lattice tile indices.
+
+Unfortunately, the indices cannot be determined with one simple formula as
+before.  Indexing requires a two-step process, a coarse step which determines a
+set of 4 tiles that contains the particle and a fine step that determines which
```
Comment (paulromano) [inline docs/source/methods/geometry.rst]: four tiles

File: docs/source/usersguide/input.rst [hunk 4]
```diff
@@ -846,6 +839,90 @@ sub-elements:
...
+    *Default*: None
+
+  :center:
+    The coordinates of the center of the lattice. If the lattice does not have
+    axial sections then only the x- and y-coordinates are specified.
+
+    *Default*: None
+
+  :pitch:
+    The pitch between lattice cells in the x- and y- (and z-) directions.
```
Comment (paulromano) [inline docs/source/usersguide/input.rst]: This description should probably be modified since y-axis doesn't make sense for hexagonal lattices.

File: src/geometry_header.F90 [hunk 5]
```diff
@@ -72,4 +148,166 @@ module geometry_header
...
+contains
+
+!===============================================================================
+
+  function valid_inds_rect(this, i_xyz) result(is_valid)
+    class(RectLattice), intent(in) :: this
+    integer,            intent(in) :: i_xyz(3)
+    logical                        :: is_valid
+
+    is_valid = (all(i_xyz > 0 .and. i_xyz <= this % n_cells))
```
Comment (paulromano) [inline src/geometry_header.F90]: No need for parentheses outside of `all`

File: src/geometry.F90 [hunk 6]
```diff
@@ -759,34 +648,47 @@ end subroutine cross_lattice
...
-!$omp threadprivate(cl, surf, lat, coord, final_coord)
+    type(LocalCoord), pointer, save :: parent_coord => null()
+!$omp threadprivate(cl, surf, lat, coord, final_coord, parent_coord)
 
     ! inialize distance to infinity (huge)
     dist = INFINITY
-    lattice_crossed = NONE
+    d_lat = INFINITY
+    d_surf = INFINITY
+    lattice_translation = (/0, 0, 0/)
```
Comment (paulromano) [inline src/geometry.F90]: I prefer bracket notation for arrays, i.e. `[0, 0, 0]`, which is equivalent to `(/0, 0, 0,/)` per the Fortran standard. Two fewer characters, and it's reminiscent of Python :) But the choice is yours.

**Comment by paulromano** [comment]
Need to update models in examples/ folder.

File: src/geometry.F90 [hunk 7]
```diff
@@ -173,170 +169,102 @@ recursive subroutine find_cell(p, found, search_cells)
...
-                     &defined outer universe.")
-              else
-                p % coord % next % universe = lat % outer
-              end if
-            end if
+        ! set particle lattice indices
+        p % coord % next% lattice   = c % fill
+        p % coord % next% lattice_x = i_xyz(1)
+        p % coord % next% lattice_y = i_xyz(2)
+        p % coord % next% lattice_z = i_xyz(3)
```
Comment (paulromano) [inline src/geometry.F90]: Very clean, I love it!

File: docs/source/methods/geometry.rst (hunk 1 above)
Comment (smharper) [inline docs/source/methods/geometry.rst]: It looks like \left and \right work fine.  I've switched over to those now.

File: docs/source/usersguide/input.rst (hunk 4 above)
Comment (smharper) [inline docs/source/usersguide/input.rst]: I made the pitch descriptions more specific.  Let me know if you think they are too wordy now.

File: src/geometry.F90 (hunk 6 above)
Comment (smharper) [inline src/geometry.F90]: I'm indifferent so I'll defer to your preference.

File: src/input_xml.F90 [hunk 8]
```diff
@@ -1298,24 +1294,42 @@ subroutine read_geometry_xml()
...
-          get_arraysize_double(node_lat, "width")) then
-        call fatal_error("Number of entries on <width> must be the same as &
-             &the number of entries on <lower_left>.")
-      end if
+      ! Read lattice pitches.
+      ! TODO: Remove this deprecation warning in a future release.
+      if (check_for_node(node_lat, "width")) then
+        call warning("The use of 'width' is deprecated and will be disallowed &
+             &in a future release.  Use 'pitch' instead.  The utility openmc/&
+             &src/utils/update_lattices.py can be used to automatically update &
```
Comment (paulromano) [inline src/input_xml.F90]: update_inputs.py

File: src/geometry.F90 [hunk 9]
```diff
@@ -759,34 +648,47 @@ end subroutine cross_lattice
...
-!$omp threadprivate(cl, surf, lat, coord, final_coord)
+    type(LocalCoord), pointer, save :: parent_coord => null()
+!$omp threadprivate(cl, surf, lat, coord, final_coord, parent_coord)
 
     ! inialize distance to infinity (huge)
     dist = INFINITY
-    lattice_crossed = NONE
+    d_lat = INFINITY
+    d_surf = INFINITY
+    lattice_translation = [0, 0, 0]
```
Comment (paulromano) [inline src/geometry.F90]: Also, whenever you assign an array like this and the left-hand side is already allocated, it would be wise to write it as `lattice_translation(:) = [0, 0, 0]`. The Fortran 2003 standard actually allows you to reallocate an array by assignment, so technically that could happen. Of course, a smart compiler would avoid it, but to be on the safe side, it's good to be specific.

Comment (smharper) [inline src/geometry.F90]: Good to know.  It should be fixed now.

File: src/error.F90 [hunk 10]
```diff
@@ -18,24 +18,36 @@ module error
...
+    logical, optional, intent(in) :: deprecation ! is a depreciation warning
 
-    ! Write warning at beginning
-    write(ERROR_UNIT, fmt='(1X,A)', advance='no') 'WARNING: '
+    integer                   :: i_start      ! starting position
+    integer                   :: i_end        ! ending position
+    integer                   :: line_wrap    ! length of line
+    integer                   :: length       ! length of message
+    integer                   :: indent       ! length of indentation
+    character(:), allocatable :: message      ! input message with a prefix
```
Comment (paulromano) [inline src/error.F90]: Unfortunately, allocatable characters break building with gfortran 4.6. Normally, I wouldn't care much, but since we're moving to Travis CI for testing and their VMs still use Ubuntu 12.04 (which in turn uses gcc 4.6), this would prevent building there. Would you mind re-coding this to remove allocatable characters?

File: src/geometry.F90 [hunk 11]
```diff
@@ -770,34 +649,48 @@ end subroutine cross_lattice
...
+    real(8) :: r                 ! radius for quadratic surfaces
+    real(8) :: tmp               ! dot product of surface normal with direction
+    real(8) :: a,b,c,k           ! quadratic equation coefficients
+    real(8) :: quad              ! discriminant of quadratic equation
+    logical :: on_surface        ! is particle on surface?
     type(Cell),       pointer, save :: cl => null()
     type(Surface),    pointer, save :: surf => null()
-    type(Lattice),    pointer, save :: lat => null()
+    class(Lattice),   pointer, save :: lat => null()
     type(LocalCoord), pointer, save :: coord => null()
```
Comment (paulromano) [inline src/geometry.F90]: This goes back a long way to when threading was first being implemented. I think the explanation is as follows: if a variable is initialized at the time it's declared, it implicitly has the save attribute. By virtue of being saved, it is treated as shared data, and therefore if you want it to be threadprivate you must explicitly declare it as such. Conversely, a threadprivate variable must be declared with the save attribute. We may be able to get rid of all the threadprivate and save declarations by not initializing pointers in procedures.

File: src/error.F90 (hunk 10 above)
Comment (smharper) [inline src/error.F90]: I don't think reworking `warning` is worth the effort right now.  If it's okay with you, I'd rather just revert to the develop version of `warning`.

**Comment by paulromano** [comment]
Thanks again @smharper!

File: docs/source/usersguide/input.rst [hunk 12]
```diff
@@ -846,6 +841,92 @@ sub-elements:
...
+    *Default*: None
+
+Here is an example of a properly defined 2d hexagonal lattice:
+
+.. code-block:: xml
+
+    <hex_lattice id="10" n_rings="3" outer="1">
+        <center> 0.0 0.0 </center>
+        <pitch> 1.0 </pitch>
+        <universes>
```
Comment (wbinventor) [inline docs/source/usersguide/input.rst]: I know this PR has been posted, but is it a typo that this example hex lattice specification does not include the "n_rings" element? Based on the description of "n_rings" above, I believe this is a required element.

Comment (smharper) [inline docs/source/usersguide/input.rst]: The "n_rings" is present as an attribute in this case.

Comment (wbinventor) [inline docs/source/usersguide/input.rst]: I totally missed that. Sorry for the confusion!

Comment (smharper) [inline docs/source/usersguide/input.rst]: No worries
